File: basic/machine-learning/linear-regression/12 linear-regression.rb
def linear_regression(x, y)
  n = x.length
  sum_x = x.sum
  sum_y = y.sum
  sum_xy = x.zip(y).map { |a, b| a * b }.sum
  sum_x2 = x.map { |a| a * a }.sum

  m = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x)
  b = (sum_y - m * sum_x) / n

  return m, b
end

x = [1, 2, 3, 4, 5]
y = [2, 3, 5, 7, 11]
m, b = linear_regression(x, y)
puts "y = #{m.round(2)}x + #{b.round(2)}"

File: basic/machine-learning/linear-regression/12 linear-regression.js
function linearRegression(x, y) {
  const n = x.length;
  let sumX = 0, sumY = 0, sumXY = 0, sumX2 = 0;

  for (let i = 0; i < n; i++) {
      sumX += x[i];
      sumY += y[i];
      sumXY += x[i] * y[i];
      sumX2 += x[i] * x[i];
  }

  const m = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);
  const b = (sumY - m * sumX) / n;

  return { m, b };
}

const x = [1, 2, 3, 4, 5];
const y = [2, 3, 5, 7, 11];
const { m, b } = linearRegression(x, y);
console.log(`y = ${m.toFixed(2)}x + ${b.toFixed(2)}`);